heartbeat/monitors/wrappers/monitorstate/monitorstate_test.go

```diff
@@ -0,0 +1,96 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package monitorstate
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/beats/v7/heartbeat/monitors/stdfields"
+)
+
+func TestRecordingAndFlapping(t *testing.T) {
+	ms := newMonitorState(TestSf, StatusUp, 0, true)
+	recordFlappingSeries(TestSf, ms)
+	require.Equal(t, StatusFlapping, ms.Status)
+	require.Equal(t, FlappingThreshold+1, ms.Checks)
+	require.Equal(t, ms.Up+ms.Down, ms.Checks)
+
+	recordStableSeries(TestSf, ms, FlappingThreshold*2, StatusDown)
+	require.Equal(t, StatusDown, ms.Status)
+	// The count should be FlappingThreshold+1 since we used double the threshold before
+	// This is because we have one full threshold of stable checks, as well as the final check that
+	// flipped us out of the threshold, which goes toward the new state.
+	requireMSCounts(t, ms, 0, FlappingThreshold+1)
+	require.Nil(t, ms.Ends, "expected nil ends after a stable series")
+
+	// Since we're now in a stable state a single up check should create a new state from a stable one
+	ms.recordCheck(TestSf, StatusUp)
+	require.Equal(t, StatusUp, ms.Status)
+	requireMSCounts(t, ms, 1, 0)
+}
+
+func TestDuration(t *testing.T) {
+	ms := newMonitorState(TestSf, StatusUp, 0, true)
+	ms.recordCheck(TestSf, StatusUp)
+	time.Sleep(time.Millisecond * 10)
+	ms.recordCheck(TestSf, StatusUp)
+	// Pretty forgiving upper bound to account for flaky CI
+	require.True(t, ms.DurationMs > 9 && ms.DurationMs < 300, "Expected duration to be ~10ms, got %d", ms.DurationMs)
```

Comment:
nit: we could remove the 300ms upper bound here.I see no harm in that.

Reply:
My worry is that there could be a bug in how we calculate the duration in the future. Let me up it to say 900 though, that's very forgiving.